### PR TITLE
[Checkpointer] Remove the dependencies on PyTorch distributed state_dict APIs

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #3638
* __->__ #3623

**Summary:**
Same as https://github.com/pytorch/torchtitan/pull/2441/ but with new implementation to be compatiblewith the latest OptimizerContainer

**Verification:**
  | Check | Result |
  |---|---|
  | Checkpoint compat, FSDP (old saves -> new loads, resume to step 100) | PASS - bit-identical all 100
  steps |
  | Checkpoint compat, PP=2 (old saves -> new loads) | PASS - bit-identical |
  | loss_compare, FSDP (old vs new, 100 steps, seed checkpoint) | PASS - identical loss |
  | CPU unit tests (lr_scheduler, optimizer_param_groups, state_dict_keys, checkpoint) | 51 passed |
  | Full unit suite | 411 passed (2 unrelated tokenizer-download failures) |
  | Lint (ufmt, flake8, pydoclint) | pass |

### CODE REVIEW

File: torchtitan/overrides/fused_swiglu.py
Comment (tianyu-l) [inline torchtitan/overrides/fused_swiglu.py:1]: Now that we are not using DCP APIs, could you prompt claude to add the `_save_to_state_dict` / ``_load_from_state_dict` hooks and see if it just solves https://github.com/pytorch/torchtitan/issues/3569

File: torchtitan/components/lr_scheduler.py [hunk 1]
```diff
@@ -201,6 +201,10 @@ def __init__(self, optimizers: OptimizersContainer, lr_lambda: Callable) -> None
         ), "Must have at least one optimizer to create LRScheduler"
 
         self.schedulers = [LambdaLR(optimizer, lr_lambda) for optimizer in optimizers]
+        # Per-optimizer regex patterns (aligned with self.schedulers), used as lr
+        # metric labels. Sourced from the container so patterns stay off the
+        # optimizer param groups and out of the saved state dict.
+        self._param_group_patterns = optimizers._param_group_patterns
```
Comment (tianyu-l) [inline torchtitan/components/lr_scheduler.py]: can we not add this simply for logging purpose? I think it's available in `optimizers.config`?

File: torchtitan/components/optimizer.py [hunk 2]
```diff
@@ -119,6 +119,11 @@ class Config(Configurable.Config):
 
     optimizers: list[T]
     model_parts: list[nn.Module]
+    # Regex patterns per optimizer (aligned with ``optimizers``); each entry is
+    # the list of patterns for that optimizer's param groups. Kept here, off the
+    # optimizer param groups, so they feed logging and lr metrics without leaking
+    # into the saved optimizer state dict.
+    _param_group_patterns: list[list[str]]
```
Comment (tianyu-l) [inline torchtitan/components/optimizer.py]: It seems it's also only for logging. Let's remove for now, or we can move the logging into `_build_param_groups` where we still have access to `patterns`.

**Review by meetv18 — APPROVED** [review]
From checkpoint POV this change makes sense to me. We should def upstream some of this! for e.g  to make things easier with optimizer SD handling.

File: torchtitan/components/lr_scheduler.py (hunk 1 above)
Comment (fegin) [reply torchtitan/components/lr_scheduler.py]: We usually don't keep `config`. I remove this but the downside is that the log now loses the pattern and is now replaced with index only.

File: torchtitan/components/optimizer.py (hunk 2 above)
Comment (fegin) [reply torchtitan/components/optimizer.py]: This can be removed. But lr_scheduler, which is a different Class now loses the patter access.

File: torchtitan/overrides/fused_swiglu.py
Comment (fegin) [reply torchtitan/overrides/fused_swiglu.py:1]: We don't need to modify `_save_to_state_dict` and `_load_from_state_dict`. nn.Module provides the APIs to register hook. https://github.com/pytorch/torchtitan/pull/3638. I haven't reviewed the details yet, but the code looks legit.

**Review by fegin — COMMENTED** [review]
We need to decide whether we want to keep pattern for lr_scheduler logging purpose.

**Review by tianyu-l — APPROVED** [review]
